### PR TITLE
SONAR-6303 WS to compare quality profiles

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

**Comment by GithubIntegration** [comment]
SonarQube analysis reported 3 new issues:
* 1 major
* 1 minor
* 1 info

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.qualityprofile.db.QualityProfileDto;
+import org.sonar.core.util.NonNullInputFunction;
+import org.sonar.server.db.DbClient;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+public class QProfileComparator implements ServerComponent {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: "Comparator" makes me thing this class implements the Comparator interface. You could maybe rename it "QProfileComparison"

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Where are the tests ?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      writeRule(json, key, rulesByKey);
+
+      json.name("left").beginObject();
+      json.prop("severity", value.leftSeverity());
+      json.name("params").beginObject();
+      for (Entry<String, ValueDifference<String>> valueDiff : value.paramDifference().entriesDiffering().entrySet()) {
+        json.prop(valueDiff.getKey(), valueDiff.getValue().leftValue());
+      }
+      for (Entry<String, String> valueDiff : value.paramDifference().entriesOnlyOnLeft().entrySet()) {
+        json.prop(valueDiff.getKey(), valueDiff.getValue());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java]: No test is covering this line

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      json.endObject();
+
+      json.name("right").beginObject();
+      json.prop("severity", value.rightSeverity());
+      json.name("params").beginObject();
+      for (Entry<String, ValueDifference<String>> valueDiff : value.paramDifference().entriesDiffering().entrySet()) {
+        json.prop(valueDiff.getKey(), valueDiff.getValue().rightValue());
+      }
+      for (Entry<String, String> valueDiff : value.paramDifference().entriesOnlyOnRight().entrySet()) {
+        json.prop(valueDiff.getKey(), valueDiff.getValue());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java]: No test is covering this line

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: I would test : 
- 2 identical profiles
- 2 completly different profiles
- 2 profiles having same params (but inserted in different order)
...

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        result.inLeft.put(ruleKey, leftActiveRulesByRuleKey.get(ruleKey));
+      } else {
+        compare(leftActiveRulesByRuleKey.get(ruleKey), rightActiveRulesByRuleKey.get(ruleKey), result);
+      }
+    }
+  }
+
+  private void compare(ActiveRule leftRule, ActiveRule rightRule, QProfileComparisonResult result) {
+    RuleKey key = leftRule.key().ruleKey();
+    if (leftRule.params().equals(rightRule.params()) && leftRule.severity().equals(rightRule.severity())) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Are you sure comparison of hashmap with equals() is working when order of param is different ?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  public static class ActiveRuleDiff {
+    private RuleKey key;
+    private String leftSeverity;
+    private String rightSeverity;
+    private MapDifference<String, String> paramDifference;
+
+    public RuleKey key() {
+      return key;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Is is really used ? It's not covered by test

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    }
+  }
+
+  private void compare(ActiveRule leftRule, ActiveRule rightRule, QProfileComparisonResult result) {
+    RuleKey key = leftRule.key().ruleKey();
+    if (leftRule.params().equals(rightRule.params()) && leftRule.severity().equals(rightRule.severity())) {
+      result.same.put(key, leftRule);
+    } else {
+      ActiveRuleDiff diff = new ActiveRuleDiff();
+      diff.key = key;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Why storing only the left rule key ? But as I can see, this field is never read.

**Comment by julienlancelot** [comment]
Please apply my feedback

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 4 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: java.util.Map is not ordered, how could insert order change equality?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 5 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Good point.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 6 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: If you follow the flow, here we are comparing 2 active rules which concern the same rule. I will rename the method to highlight the intent.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 4 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Good point !

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java (hunk 6 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileComparator.java]: Ok. But still, is this variable still needed ?

**Comment by jblievremont** [comment]
Feedback applied, thanks!

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+      .prop("name", rulesByKey.get(key).name())
+      .prop("pluginKey", repositoryKey);
+
+    Repository repo = ruleRepositories.repository(repositoryKey);
+    if (repo != null) {
+      String languageKey = repo.getLanguage();
+      Language language = languages.get(languageKey);
+
+      json.prop("pluginName", repo.getName());
+      json.prop("languageKey", languageKey);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileCompareAction.java]: I'm wondering, does it make sense to compare 2 profiles from different language ? I know that in rails it was possible, but I'm use it makes sense to keep this feature...

**Comment by julienlancelot** [comment]
LGTM !

**Pull request closed without merging**